astropy/stats/sigma_clipping.py

```diff
@@ -378,7 +378,9 @@ def _sigmaclip_withaxis(self, data, axis=None, masked=True,
         if masked:
             # create an output masked array
             if copy:
-                filtered_data = np.ma.masked_invalid(filtered_data)
+                filtered_data = np.ma.masked_invalid(filtered_data, copy=False)
+                filtered_data = np.ma.MaskedArray(data, filtered_data.mask,
+                                                  copy=True)
```

Comment:
```suggestion
                filtered_data = np.ma.MaskedArray(data, ~np.isfinite(filtered_data),
                                                  copy=True)
```

This is ~10x faster (tested with a 1k x 1k array) than using `np.ma.masked_invalid`.

Reply:
OK. But surely this is going to be true for the other instances of `masked_invalid` as well? The 10x difference must be in the existing calls, rather than the line I added, so perhaps we should change all of them? As far as I can see, the only differences are that `masked_invalid` uses `where` and that it ORs the result with any existing mask (if I understand the docs). The latter thing might need doing separately at L346 (but that presumably can't account for most of the 10x). The other 2 instances could probably just be replaced in the same way as above?

Reply:
`masked_invalid` is using `np.isfinite` so I'm surprised by the difference. And one thing that `masked_invalid` does additionally is to combine the masks if the input is a masked array. I think we need to keep that, as here `data` may be a masked array ?

Reply:
Me too. Maybe constructing the indices in `masked_where` is fiddly (and I think unnecessary)? I think we need to keep the mask combination as you say at L346, but after that the data get converted to an `ndarray` for intermediate processing.

Reply:
I should really just time it when I get a minute (but I have to finish other things this week).

Reply:
Yes actually the masked values in the input data have been set to NaN, so there is no need to combine with the input mask.

Reply:
` filtered_data = np.ma.masked_invalid(filtered_data, copy=False)` above was being used *only* to get a new mask (that is then used in the next line) -- effectively it was being used only to do `~np.isfinite(filtered_data)` (`filtered_data` is a plain ndarray), so I suggested that we just use `np.isfinite` directly to get the new mask.

My timing test was a simple comparison of only `np.ma.masked_invalid` vs. `np.isfinite` on a 1k x 1k array.  The former took 2.1 ms and the latter 311 us (~7x slower).  The difference is likely due to the creation/initialization of a `MaskedArray` object.

Reply:
Oh! I think I understand what you're getting at now. I thought you were saying that `MaskedArray` with `np.isfinite` was 10x as fast as `masked_invalid` plus `MaskedArray`, which would have implied that `masked_invalid` will be slow everywhere. But you're just saying that, assuming there's always a `MaskedArray`, adding a `masked_invalid` takes 10x as long as adding an `isfinite` to it (so the 2 lines probably take ~twice as long as your suggested one). OK, I'll just accept your suggestion then...